internal/controller/csiaddons/persistentvolumeclaim_controller.go

```diff
@@ -210,97 +215,46 @@ func (r *PersistentVolumeClaimReconciler) checkDriverSupportCapability(
 
-	// If the schedule is found, check whether driver supports the
-	// space reclamation using annotation on namespace and registered driver
-	// capability for decision on requeue.
-	if scheduleFound {
-		// We cannot have a generic solution for all CSI drivers to get the driver
-		// name from PV and check if driver supports reclaimspace/keyrotation or not and
-		// requeue the request if the driver is not registered in the connection
-		// pool. This can put the controller in a requeue loop. Hence we are
-		// reading the driver name from the namespace annotation and checking if
-		// the driver is registered in the connection pool and if not we are not
-		// requeuing the request.
-		// Depending on requeue value, it will return ErrorConnNotFoundRequeueNeeded.
-		switch annotationKey {
-		case krcJobScheduleTimeAnnotation:
-			requeue, keyRotationSupported := r.checkDriverSupportCapability(logger, ns.Annotations, driverName, keyRotationOp)
-			if keyRotationSupported {
-				return schedule, nil
-			}
-			if requeue {
-				return "", ErrConnNotFoundRequeueNeeded
-			}
-		case rsCronJobScheduleTimeAnnotation:
-			requeue, supportReclaimspace := r.checkDriverSupportCapability(logger, ns.Annotations, driverName, relciamSpaceOp)
-			if supportReclaimspace {
-				// if driver supports space reclamation,
-				// return schedule from ns annotation.
-				return schedule, nil
-			}
-			if requeue {
-				// The request needs to be requeued for checking
-				// driver support again.
-				return "", ErrConnNotFoundRequeueNeeded
-			}
-		default:
-			logger.Info("Unknown annotation key", "AnnotationKey", annotationKey)
-			return "", fmt.Errorf("unknown annotation key: %s", annotationKey)
+	if r.SchedulePrecedence == util.ScheduleSCOnly {
+		if schedule = r.getScheduleFromSC(ctx, pvc, logger, annotationKey); schedule != "" {
+			return schedule, nil
 		}
-	}
-
-	// For static provisioned PVs, StorageClassName is nil or empty.
-	if pvc.Spec.StorageClassName == nil || len(*pvc.Spec.StorageClassName) == 0 {
-		logger.Info("StorageClassName is empty")
 		return "", ErrScheduleNotFound
 	}
-	storageClassName := *pvc.Spec.StorageClassName
 
-	// check for storageclass schedule annotation.
-	sc := &storagev1.StorageClass{}
-	err = r.Client.Get(ctx, types.NamespacedName{Name: storageClassName}, sc)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			logger.Error(err, "StorageClass not found", "StorageClass", storageClassName)
-			return "", ErrScheduleNotFound
-		}
+	// Check on PVC
+	if schedule = r.getScheduleFromPVC(pvc, logger, annotationKey); schedule != "" {
+		return schedule, nil
+	}
 
-		logger.Error(err, "Failed to get StorageClass", "StorageClass", storageClassName)
+	// Check on NS, might get ErrConnNotFoundRequeueNeeded
+	// If so, return the error
+	if schedule, err = r.getScheduleFromNS(ctx, pvc, logger, driverName, annotationKey); schedule != "" {
+		return schedule, nil
+	}
+	if errors.Is(err, ErrConnNotFoundRequeueNeeded) {
```

Comment:
you should proceed to check schedule on SC only when you get ErrScheduleNotFound here right ? 

you need to error out for ConnNotFound + errors other than ErrScheduleNotFound ?

Reply:
Right, in cases where the schedule in present only on the NS and we are unable to fetch the NS.